feat: Editorial review: Add docs for notRestoredReasons (#32920)
mdn/content: files/en-us/web/api/performance_api/reporting_backforward_cache_not_restored_reasons/index.md

---

> title: Reporting back/forward cache not restored reasons

Author: This seems a bit too vague/general to me — this is specifically about monitoring reasons a document was blocked from using the bfcache. I thought about it a bit, and updated it to "Monitoring back/forward cache blocking reasons", which is slightly shorter, and a bit less wordy ... maybe?

slug: Web/API/Performance_API/Reporting_backforward_cache_not_restored_reasons
page-type: web-api-overview
status:
  - experimental
browser-compat: api.PerformanceNavigationTiming.notRestoredReasons
---

{{DefaultAPISidebar("Performance API")}}{{SeeCompatTable}}

The {{domxref("PerformanceNavigationTiming.notRestoredReasons")}} property reports information on why the current document was blocked from using the [back/forward cache (bfcache)](https://web.dev/bfcache/) on navigation. Developers can use this information to identify pages that need updates to make them bfcache-compatible, thereby improving site performance.

> ## Concepts and usage

Author: I've used "## Back/forward cache (bfcache)"

Modern browsers provide an optimization feature for history navigation called the [back/forward cache (bfcache)](https://web.dev/bfcache/). This enables an instant loading experience when users go back to a page they have already visited. Pages can be blocked from entering the bfcache or get evicted while in the bfcache for different reasons, some required by a specification and some specific to browser implementations.

> To enable monitoring bfcache blocking in the field the [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) class includes a `notRestoredReasons` property. This returns an object containing related information on the top-level frame and all {{htmlelement("iframe")}}s present in the document:

Author: I updated it to "To enable monitoring bfcache blocking reasons, the..."

- Reasons why bfcache usage was blocked.
- Details such as frame `id` and `name`, to help identify `<iframe>`s in the HTML.

> **Note:** There is also a [test in Chrome dev tools](https://web.dev/bfcache/#test) that reports on whether your pages are cacheable.

> ## Examples

Author: I ended up using "## Logging bfcache blocking reasons"

A [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) instance can be obtained from features such as [`Performance.getEntriesByType()`](/en-US/docs/Web/API/Performance/getEntriesByType) and [`PerformanceObserver`](/en-US/docs/Web/API/PerformanceObserver).

For example, you could invoke the following function to return all `PerformanceNavigationTiming` objects currently present in the performance timeline and log their `notRestoredReasons`:

```js
function returnNRR() {

>   const navEntries = performance.getEntriesByType("navigation");

Author: That makes sense. I've added a `PerformanceObserver` example and restructured the section a bit to suit.

- {{domxref("NotRestoredReasons.url", "url")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the URL of the navigated page or `<iframe>`. If the document is in a cross-origin `<iframe>`, `url` will return `null`.

### Reporting bfcache blocking in same-origin `<iframe>`s

When a page has same-origin `<iframe>`s embedded, the returned `notRestoredReasons` value will contain an array of objects inside the `children` property representing the blocking reasons related to each embedded frame.

For example:

```js
{
  children: [
    {
      children: [],
      id: "iframe-id",
      name: "iframe-name",
      reasons: [],
      src: "./index.html",
      url: "https://www.example.com/iframe-examples.html"
    },
    {
      children: [],
      id: "iframe-id2",
      name: "iframe-name2",
      reasons: [
        { "reason": "unload-listener" }
      ],
      src: "./unload-examples.html",
      url: "https://www.example.com/unload-examples.html"
    },
  ],
  id: null,
  name: null,
  reasons: [],
  src: null,
  url:"https://www.example.com"
}
```

### Reporting bfcache blocking in cross-origin `<iframe>`s

> When a page has cross-origin frames embedded, we limit the amount of information shared about them to avoid leaking cross-origin information. We only include information that the outer page already knows, and whether the cross-origin subtree caused bfcache blocking or not. We don't include any blocking reasons or information about lower levels of the subtree (even if some sub-levels are same-origin).

Reviewer: Who is "we"? I guess it is not the web developer. This "we" seems to refer to the implementation or the specification and that is irrelevant to the web developer.
 
```suggestion
When a page has cross-origin frames embedded, the amount of information shared about them is limited to avoid leaking cross-origin information. Only information that the outer page already knows is included, and whether the cross-origin subtree caused bfcache blocking or not. No blocking reasons or information about lower levels of the subtree (even if some sub-levels are same-origin) are included.
```
